File: integrations/pi/plugins/powercontext/src/commands.ts
/*
 * Copyright (c) 2026 OceanBase.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { ExtensionAPI } from '@earendil-works/pi-coding-agent'
import {
  confirmDurableWrite,
  invokeOperation,
  invokeScopedOperation,
  unavailableResult,
  type ToolResult,
} from './invoke.ts'
import type { PluginRuntime } from './recall.ts'

type CommandContext = {
  cwd: string
  hasUI: boolean
  signal?: AbortSignal
  ui: {
    confirm: (title: string, message: string) => Promise<boolean>
    notify: (message: string, level: 'info' | 'error') => void
  }
}

function format(result: ToolResult | Record<string, unknown>): string {
  return JSON.stringify(result, null, 2)
}

function report(context: CommandContext, message: string, isError = false): void {
  if (context.hasUI) {
    context.ui.notify(message, isError ? 'error' : 'info')
    return
  }
  console.log(message)
}

async function handleRemember(runtime: PluginRuntime, context: CommandContext, text: string): Promise<ToolResult> {
  if (!text) return { ok: false, code: 'invalid_request', message: 'Usage: /pc remember <text>' }
  const confirmation = await confirmDurableWrite(context, 'remember operation')
  if (confirmation) return confirmation
  return invokeScopedOperation(runtime, context, 'remember_memory', { kind: 'agent-note', text })
}

export async function handlePcCommand(
  rawInput: string,
  runtime: PluginRuntime,
  context: CommandContext,
): Promise<void> {
  const tokens = rawInput.trim().split(/\s+/).filter(Boolean)
  const command = tokens[0]
  if (!command) {
    try {
      const scopeId = await runtime.resolveScope(context.cwd)
      report(context, `scope=${scopeId}\nbaseUrl=${runtime.config.baseUrl}\nUse /pc doctor to check Server readiness.`)
    } catch {
      report(context, format(unavailableResult()), true)
    }
    return
  }
  if (command === 'doctor') {
    try {
      const scopeId = await runtime.resolveScope(context.cwd)
      const [live, ready] = await Promise.all([
        invokeOperation(runtime.client, 'get_liveness', {}, scopeId, context.signal),
        invokeOperation(runtime.client, 'get_readiness', {}, scopeId, context.signal),
      ])
      report(context, format({ ok: live.ok && ready.ok, live, ready }), !live.ok || !ready.ok)
    } catch {
      report(context, format(unavailableResult()), true)
    }
    return
  }
  if (command === 'search') {
    const query = tokens.slice(1).join(' ')
    const result = query
      ? await invokeScopedOperation(runtime, context, 'search_memory', { query, limit: 8, mode: 'auto' })
      : { ok: false, code: 'invalid_request', message: 'Usage: /pc search <query>' }
    report(context, format(result), !result.ok)
    return
  }
  if (command === 'remember') {
    const result = await handleRemember(runtime, context, tokens.slice(1).join(' '))
    report(context, format(result), !result.ok)
    return
  }
  if (command === 'flush') {
    const confirmation = await confirmDurableWrite(context, 'flush operation')
    const result = confirmation ?? await invokeScopedOperation(runtime, context, 'flush_memory', {})
    report(context, format(result), !result.ok)
    return
  }
  if (command === 'stats') {
    const result = await invokeScopedOperation(runtime, context, 'get_stats', {})
    report(context, format(result), !result.ok)
    return
  }
  report(
    context,
    'Unknown /pc subcommand. Try doctor, search, remember, flush, or stats.',
    true,
  )
}

export function registerCommands(pi: ExtensionAPI, runtime: PluginRuntime): void {
  pi.registerCommand('pc', {
    description: 'PowerContext status, search, and diagnostics',
    handler: async (args, context) => {
      await handlePcCommand(args, runtime, context)
    },
  })
}

File: desktop/src-tauri/src/commands.rs
/*
 * Copyright (c) 2026 OceanBase.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

use crate::{
    connections::{
        profiles::ProfileInput,
        session::{ConnectionManager, DesktopState},
    },
    error::SafeError,
    ipc::authorize_window,
    transport::{
        ApiFailure,
        wire::{ScopeDescriptor, ScopePage},
    },
};
pub struct HostState {
    pub manager: Result<ConnectionManager, SafeError>,
}
fn manager<'a, R: tauri::Runtime>(
    window: &tauri::WebviewWindow<R>,
    state: &'a HostState,
) -> Result<&'a ConnectionManager, ApiFailure> {
    authorize_window(window.label())?;
    state.manager.as_ref().map_err(|e| ApiFailure::before(*e))
}

#[tauri::command]
pub fn desktop_state<R: tauri::Runtime>(
    window: tauri::WebviewWindow<R>,
    state: tauri::State<'_, HostState>,
) -> Result<DesktopState, ApiFailure> {
    manager(&window, &state)?.state().map_err(Into::into)
}

#[tauri::command]
pub fn save_profile<R: tauri::Runtime>(
    window: tauri::WebviewWindow<R>,
    state: tauri::State<'_, HostState>,
    input: ProfileInput,
) -> Result<DesktopState, ApiFailure> {
    manager(&window, &state)?
        .save_profile(input)
        .map_err(Into::into)
}

#[tauri::command]
pub fn remove_profile<R: tauri::Runtime>(
    window: tauri::WebviewWindow<R>,
    state: tauri::State<'_, HostState>,
    id: String,
    revision: u32,
) -> Result<DesktopState, ApiFailure> {
    manager(&window, &state)?
        .remove_profile(&id, revision)
        .map_err(Into::into)
}

#[tauri::command]
pub async fn check_connection<R: tauri::Runtime>(
    window: tauri::WebviewWindow<R>,
    state: tauri::State<'_, HostState>,
    id: String,
    activate: bool,
) -> Result<DesktopState, ApiFailure> {
    manager(&window, &state)?.check(&id, activate).await
}

#[tauri::command]
pub fn disconnect<R: tauri::Runtime>(
    window: tauri::WebviewWindow<R>,
    state: tauri::State<'_, HostState>,
) -> Result<DesktopState, ApiFailure> {
    manager(&window, &state)?.disconnect().map_err(Into::into)
}

#[tauri::command]
pub fn invalidate_profile<R: tauri::Runtime>(
    window: tauri::WebviewWindow<R>,
    state: tauri::State<'_, HostState>,
    id: String,
) -> Result<DesktopState, ApiFailure> {
    manager(&window, &state)?
        .invalidate_profile(&id)
        .map_err(Into::into)
}

#[tauri::command]
pub async fn list_scopes<R: tauri::Runtime>(
    window: tauri::WebviewWindow<R>,
    state: tauri::State<'_, HostState>,
    generation: u32,
    query: String,
    cursor: Option<String>,
) -> Result<ScopePage, ApiFailure> {
    manager(&window, &state)?
        .scopes(generation, &query, cursor.as_deref())
        .await
}

#[tauri::command]
pub async fn default_scope<R: tauri::Runtime>(
    window: tauri::WebviewWindow<R>,
    state: tauri::State<'_, HostState>,
    generation: u32,
) -> Result<ScopeDescriptor, ApiFailure> {
    manager(&window, &state)?.default_scope(generation).await
}

#[tauri::command]
pub async fn select_scope<R: tauri::Runtime>(
    window: tauri::WebviewWindow<R>,
    state: tauri::State<'_, HostState>,
    generation: u32,
    id: String,
) -> Result<DesktopState, ApiFailure> {
    manager(&window, &state)?
        .select_scope(generation, &id)
        .await
}

#[tauri::command]
pub fn cancel_scope_reads<R: tauri::Runtime>(
    window: tauri::WebviewWindow<R>,
    state: tauri::State<'_, HostState>,
    generation: u32,
) -> Result<(), ApiFailure> {
    manager(&window, &state)?
        .cancel_scope_reads(generation)
        .map_err(Into::into)
}

#[tauri::command]
pub async fn local_diagnostics<R: tauri::Runtime>(
    window: tauri::WebviewWindow<R>,
    state: tauri::State<'_, crate::diagnostics::DiagnosticHost>,
    kind: crate::diagnostics::DiagnosticKind,
) -> Result<crate::diagnostics::DiagnosticReport, SafeError> {
    authorize_window(window.label())?;
    #[cfg(windows)]
    {
        state
            .local
            .as_ref()
            .ok_or(SafeError::NotFound)?
            .run(kind)
            .await
    }
    #[cfg(not(windows))]
    {
        let _ = (state, kind);
        Err(SafeError::NotFound)
    }
}

#[tauri::command]
pub async fn remember_memory<R: tauri::Runtime>(
    window: tauri::WebviewWindow<R>,
    state: tauri::State<'_, HostState>,
    generation: u32,
    text: String,
) -> Result<crate::connections::session::WriteOutcome, ApiFailure> {
    manager(&window, &state)?.remember(generation, &text).await
}
#[tauri::command]
pub async fn search_memory<R: tauri::Runtime>(
    window: tauri::WebviewWindow<R>,
    state: tauri::State<'_, HostState>,
    generation: u32,
    query: String,
) -> Result<crate::transport::wire::SearchMemoryResponse, ApiFailure> {
    manager(&window, &state)?
        .search_memory(generation, &query)
        .await
}
#[tauri::command]
pub async fn memory_entry<R: tauri::Runtime>(
    window: tauri::WebviewWindow<R>,
    state: tauri::State<'_, HostState>,
    generation: u32,
    citation: crate::transport::wire::MemoryCitation,
) -> Result<crate::transport::wire::MemoryEntry, ApiFailure> {
    manager(&window, &state)?
        .memory_entry(generation, &citation)
        .await
}
#[tauri::command]
pub fn cancel_memory_reads<R: tauri::Runtime>(
    window: tauri::WebviewWindow<R>,
    state: tauri::State<'_, HostState>,
    generation: u32,
) -> Result<(), ApiFailure> {
    manager(&window, &state)?
        .cancel_memory_reads(generation)
        .map_err(Into::into)
}
